File: cli/polyaxon/schemas/ops/parallel/metrics.py
#!/usr/bin/python
#
# Copyright 2019 Polyaxon, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# -*- coding: utf-8 -*-
from __future__ import absolute_import, division, print_function

from marshmallow import fields, validate

from polyaxon.schemas.base import BaseConfig, BaseSchema


class Optimization(object):
    MAXIMIZE = "maximize"
    MINIMIZE = "minimize"

    MAXIMIZE_VALUES = [MAXIMIZE, MAXIMIZE.upper(), MAXIMIZE.capitalize()]
    MINIMIZE_VALUES = [MINIMIZE, MINIMIZE.upper(), MINIMIZE.capitalize()]

    VALUES = MAXIMIZE_VALUES + MINIMIZE_VALUES

    @classmethod
    def maximize(cls, value):
        return value in cls.MAXIMIZE_VALUES

    @classmethod
    def minimize(cls, value):
        return value in cls.MINIMIZE_VALUES


class SearchMetricSchema(BaseSchema):
    name = fields.Str()
    optimization = fields.Str(
        allow_none=True, validate=validate.OneOf(Optimization.VALUES)
    )

    @staticmethod
    def schema_config():
        return SearchMetricConfig


class SearchMetricConfig(BaseConfig):
    SCHEMA = SearchMetricSchema
    IDENTIFIER = "search_metric"

    def __init__(self, name, optimization=Optimization.MAXIMIZE):
        self.name = name
        self.optimization = optimization

File: operator/controllers/plugins/metrics.go
/*
Copyright 2018-2021 Polyaxon, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package plugins

import (
	"context"

	"sigs.k8s.io/controller-runtime/pkg/client"
	"sigs.k8s.io/controller-runtime/pkg/metrics"

	"github.com/prometheus/client_golang/prometheus"

	operationv1 "github.com/polyaxon/polyaxon/operator/api/v1"
)

// Metrics exposed by the operator
type Metrics struct {
	cli                     client.Client
	OperationsRunning       *prometheus.GaugeVec
	OperationsCreated       *prometheus.CounterVec
	OperationsFailedCreated *prometheus.CounterVec
}

// NewMetrics prometheus initializer
func NewMetrics(cli client.Client) *Metrics {
	m := &Metrics{
		cli: cli,
		OperationsRunning: prometheus.NewGaugeVec(
			prometheus.GaugeOpts{
				Name: "operations_running",
				Help: "Current running operations in the cluster",
			},
			[]string{"namespace"},
		),
		OperationsCreated: prometheus.NewCounterVec(
			prometheus.CounterOpts{
				Name: "operations_created",
				Help: "Total nuber of operations created",
			},
			[]string{"namespace"},
		),
		OperationsFailedCreated: prometheus.NewCounterVec(
			prometheus.CounterOpts{
				Name: "notebook_create_failed_total",
				Help: "Total nuber of operations creation failures",
			},
			[]string{"namespace"},
		),
	}

	metrics.Registry.MustRegister(m)
	return m
}

// Describe implements the prometheus.Collector interface.
func (m *Metrics) Describe(ch chan<- *prometheus.Desc) {
	m.OperationsRunning.Describe(ch)
	m.OperationsCreated.Describe(ch)
	m.OperationsFailedCreated.Describe(ch)
}

// Collect implements the prometheus.Collector interface.
func (m *Metrics) Collect(ch chan<- prometheus.Metric) {
	m.scrape()
	m.OperationsRunning.Collect(ch)
	m.OperationsCreated.Collect(ch)
	m.OperationsFailedCreated.Collect(ch)
}

// scrape gets current running operations.
func (m *Metrics) scrape() {
	operations := &operationv1.OperationList{}
	err := m.cli.List(context.TODO(), operations)
	if err != nil {
		return
	}
	stsCache := make(map[string]float64)
	for _, op := range operations.Items {
		stsCache[op.Namespace]++
	}

	for ns, op := range stsCache {
		m.OperationsRunning.WithLabelValues(ns).Set(op)
	}
}
